File: wrappers/java/src/main/java/com/hymma/licensing/LicenseClient.java
package com.hymma.licensing;

import java.lang.foreign.MemorySegment;
import java.time.Instant;
import java.util.Optional;

/**
 * High-level licensing client (check / activate / deactivate / refresh).
 *
 * <p>Thin facade over the native hymmalm core. All license logic — JWS
 * verification, status rules, machine fingerprinting, the REST client flow,
 * retry policy and clock-tamper resistance — lives in the native library so
 * every language wrapper behaves identically; this class only marshals.
 *
 * <p>One instance per thread, or serialize calls yourself — same rule as the
 * native handle it wraps.
 *
 * <pre>{@code
 * try (var client = new LicenseClient(LicenseClientOptions.builder()
 *         .productId("PRD_01KWWPEPM0N070BDAHJ7G09RGV")
 *         .clientApiKey("PUB_...")                     // never a MST_ key
 *         .jwksJson(Files.readString(Path.of("signingkeys.json")))
 *         .licensePath(home.resolve(".myapp/license.lic").toString())
 *         .build())) {
 *     LicenseStatus status = client.check();           // trial bootstrap on first run
 *     if (status == LicenseStatus.INVALID_TRIAL) {
 *         client.activate(userEnteredReceiptCode);
 *     }
 * }
 * }</pre>
 */
public final class LicenseClient implements AutoCloseable {

    private MemorySegment handle;

    public LicenseClient(LicenseClientOptions options) {
        if (options == null) {
            throw new NullPointerException("options");
        }
        handle = Native.create(options.baseUrl(), options.productId(),
                options.clientApiKey(), options.jwksJson(),
                options.format().code(), options.validDays(),
                options.machineId(), options.machineName(),
                options.licensePath());
        if (handle.address() == 0) {
            throw new IllegalArgumentException(
                    "hlm_ffi_create rejected the options (missing product/key/"
                    + "JWKS, or no hardware fingerprint available on this platform)");
        }
    }

    // -- operations ------------------------------------------------------ //

    /** Silent check; on a fresh machine this also bootstraps the trial. */
    public LicenseStatus check() {
        return guard(Native.check(handle()));
    }

    /** Attach a purchased receipt code to this machine's license. */
    public LicenseStatus activate(String receiptCode) {
        return guard(Native.activate(handle(), receiptCode));
    }

    /** Free this machine's seat (uninstall flow). */
    public LicenseStatus deactivate() {
        return guard(Native.deactivate(handle()));
    }

    /** Fetch a fresh signed license, ignoring the cache. */
    public LicenseStatus refresh() {
        return guard(Native.refresh(handle()));
    }

    // -- state ------------------------------------------------------------ //

    public LicenseStatus status() {
        return LicenseStatus.fromCode(Native.status(handle()));
    }

    public String statusName() {
        return Native.statusName(handle());
    }

    public String licenseId() {
        return Native.licenseId(handle());
    }

    public String productName() {
        return Native.productName(handle());
    }

    public String buyerEmail() {
        return Native.buyerEmail(handle());
    }

    public boolean liveMode() {
        return Native.liveMode(handle()) != 0;
    }

    public int lastHttpStatus() {
        return Native.lastHttpStatus(handle());
    }

    /** License expiry (UTC); empty when the license carries none. */
    public Optional<Instant> expires() {
        return fromUnix(Native.expires(handle()));
    }

    /** Trial end (UTC); empty when the license carries none. */
    public Optional<Instant> trialEnd() {
        return fromUnix(Native.trialEnd(handle()));
    }

    /** Receipt expiry (UTC); empty when the license carries none. */
    public Optional<Instant> receiptExpires() {
        return fromUnix(Native.receiptExpires(handle()));
    }

    /** A metadata value from the signed license ("" when absent). */
    public String metadata(String key) {
        return Native.metadata(handle(), key);
    }

    // -- statics ------------------------------------------------------------ //

    /**
     * This machine's hardware fingerprint (52-char Crockford Base32) —
     * identical to the .NET SDK's DeviceId on the same machine.
     */
    public static String machineId() {
        return Native.machineId();
    }

    /** This computer's name, as the SDK would send it. */
    public static String machineName() {
        return Native.machineName();
    }

    /**
     * Offline one-shot: verify a signed license string against
     * {@code jwksJson} and report its status at {@code now} (null uses the
     * system clock). Throws {@link LicenseException} when the string is
     * tampered or malformed, or when it is bound to a different product or
     * machine than expected.
     */
    public static LicenseStatus verify(String jws, String jwksJson,
            String expectedProductId, String expectedMachineId, Instant now) {
        long unix = now == null ? 0 : now.getEpochSecond();
        int[] status = new int[1];
        int r = Native.verify(jws, jwksJson, expectedProductId,
                expectedMachineId, unix, status);
        if (r != 0) {
            throw new LicenseException(r);
        }
        return LicenseStatus.fromCode(status[0]);
    }

    // -- lifecycle ------------------------------------------------------------ //

    @Override
    public void close() {
        if (handle != null && handle.address() != 0) {
            Native.destroy(handle);
        }
        handle = MemorySegment.NULL;
    }

    private MemorySegment handle() {
        if (handle == null || handle.address() == 0) {
            throw new IllegalStateException("LicenseClient is closed");
        }
        return handle;
    }

    private LicenseStatus guard(int err) {
        if (err != 0) {
            throw new LicenseException(err, Native.lastErrorDetail(handle()));
        }
        return status();
    }

    private static Optional<Instant> fromUnix(long seconds) {
        return seconds == Long.MIN_VALUE
                ? Optional.empty()
                : Optional.of(Instant.ofEpochSecond(seconds));
    }
}

File: wrappers/php/src/LicenseClient.php
<?php

declare(strict_types=1);

namespace Hymma\Licensing;

// Allow plain `require 'src/LicenseClient.php';` without composer.
require_once __DIR__ . '/LicenseStatus.php';
require_once __DIR__ . '/SignedFormat.php';
require_once __DIR__ . '/Native.php';
require_once __DIR__ . '/LicenseException.php';

/**
 * High-level licensing client (check / activate / deactivate / refresh).
 *
 * Thin FFI binding over the native hymmalm shared library
 * (include/hymma/hlm_ffi.h). All licensing logic — JWS verification, status
 * rules, machine fingerprinting, the REST client flow, retry policy and
 * clock-tamper resistance — lives in the native core, so every language
 * wrapper behaves identically. This class only marshals.
 *
 * One instance per thread, or serialize calls yourself — same rule as the
 * native handle it wraps.
 *
 * Typical use:
 *
 *     $client = new LicenseClient(
 *         productId: 'PRD_01KWWPEPM0N070BDAHJ7G09RGV',
 *         clientApiKey: 'PUB_...',                       // never a MST_ key
 *         jwksJson: file_get_contents('signingkeys.json'),
 *         licensePath: '/home/me/.myapp/license.lic',
 *     );
 *     $status = $client->check();                        // trial bootstrap on first run
 *     if ($status === LicenseStatus::InvalidTrial) {
 *         $client->activate($userEnteredReceiptCode);
 *     }
 */
class LicenseClient
{
    private ?\FFI\CData $handle;

    /**
     * @param array|string|null $baseUrl Either the base URL string (or null
     *        for the production server), or an options array with the same
     *        keys as the named parameters below.
     */
    public function __construct(
        array|string|null $baseUrl = null,
        ?string $productId = null,
        ?string $clientApiKey = null,
        ?string $jwksJson = null,
        SignedFormat $format = SignedFormat::Rs256,
        int $validDays = 0,
        ?string $machineId = null,
        ?string $machineName = null,
        ?string $licensePath = null,
    ) {
        if (is_array($baseUrl)) {
            $o = $baseUrl;
            $baseUrl = $o['baseUrl'] ?? null;
            $productId = $o['productId'] ?? $productId;
            $clientApiKey = $o['clientApiKey'] ?? $clientApiKey;
            $jwksJson = $o['jwksJson'] ?? $jwksJson;
            $format = $o['format'] ?? $format;
            $validDays = $o['validDays'] ?? $validDays;
            $machineId = $o['machineId'] ?? $machineId;
            $machineName = $o['machineName'] ?? $machineName;
            $licensePath = $o['licensePath'] ?? $licensePath;
        }
        if ($format instanceof SignedFormat) {
            $format = $format->value;
        }

        $handle = Native::get()->hlm_ffi_create(
            $baseUrl,
            $productId,
            $clientApiKey,
            $jwksJson,
            (int) $format,
            $validDays,
            $machineId,
            $machineName,
            $licensePath
        );
        if ($handle === null || \FFI::isNull($handle)) {
            throw new \InvalidArgumentException(
                'hlm_ffi_create rejected the options (missing product/key/'
                . 'JWKS, or no machine fingerprint available on this platform)'
            );
        }
        $this->handle = $handle;
    }

    // -- lifecycle ------------------------------------------------------ //

    public function close(): void
    {
        if ($this->handle !== null) {
            Native::get()->hlm_ffi_destroy($this->handle);
            $this->handle = null;
        }
    }

    public function __destruct()
    {
        $this->close();
    }

    private function h(): \FFI\CData
    {
        if ($this->handle === null) {
            throw new \LogicException('LicenseClient is closed');
        }

        return $this->handle;
    }

    /** @throws LicenseException when the native call returned a nonzero code */
    private function guard(int $err): LicenseStatus
    {
        if ($err !== 0) {
            $detail = Native::str(
                Native::get()->hlm_ffi_last_error_detail($this->h())
            );
            throw new LicenseException($err, $detail);
        }

        return $this->status();
    }

    // -- operations ------------------------------------------------------ //

    /** Silent check; on a fresh machine this also bootstraps the trial. */
    public function check(): LicenseStatus
    {
        return $this->guard(Native::get()->hlm_ffi_check($this->h()));
    }

    /** Attach a purchased receipt code to this machine's license. */
    public function activate(string $receiptCode): LicenseStatus
    {
        return $this->guard(Native::get()->hlm_ffi_activate($this->h(), $receiptCode));
    }

    /** Free this machine's seat (uninstall flow). */
    public function deactivate(): LicenseStatus
    {
        return $this->guard(Native::get()->hlm_ffi_deactivate($this->h()));
    }

    /** Fetch a fresh signed license, ignoring the cache. */
    public function refresh(): LicenseStatus
    {
        return $this->guard(Native::get()->hlm_ffi_refresh($this->h()));
    }

    // -- state ------------------------------------------------------------ //

    public function status(): LicenseStatus
    {
        return LicenseStatus::from(Native::get()->hlm_ffi_status($this->h()));
    }

    public function statusName(): string
    {
        return Native::str(Native::get()->hlm_ffi_status_name($this->h()));
    }

    public function licenseId(): string
    {
        return Native::str(Native::get()->hlm_ffi_license_id($this->h()));
    }

    public function productName(): string
    {
        return Native::str(Native::get()->hlm_ffi_product_name($this->h()));
    }

    public function buyerEmail(): string
    {
        return Native::str(Native::get()->hlm_ffi_buyer_email($this->h()));
    }

    public function liveMode(): bool
    {
        return Native::get()->hlm_ffi_live_mode($this->h()) !== 0;
    }

    public function expires(): ?\DateTimeImmutable
    {
        return self::fromUnix(Native::get()->hlm_ffi_expires($this->h()));
    }

    public function trialEnd(): ?\DateTimeImmutable
    {
        return self::fromUnix(Native::get()->hlm_ffi_trial_end($this->h()));
    }

    public function receiptExpires(): ?\DateTimeImmutable
    {
        return self::fromUnix(Native::get()->hlm_ffi_receipt_expires($this->h()));
    }

    public function metadata(string $key): string
    {
        return Native::str(Native::get()->hlm_ffi_metadata($this->h(), $key));
    }

    public function lastHttpStatus(): int
    {
        return Native::get()->hlm_ffi_last_http_status($this->h());
    }

    // -- stateless helpers ------------------------------------------------ //

    /**
     * This machine's hardware fingerprint (52-char Crockford Base32) —
     * identical to the .NET SDK's DeviceId on the same machine.
     */
    public static function machineId(): string
    {
        $ffi = Native::get();
        $buf = $ffi->new('char[64]');
        $r = $ffi->hlm_ffi_machine_id($buf, 64);
        if ($r !== 0) {
            throw new LicenseException($r);
        }

        return \FFI::string($buf);
    }

    /** This computer's name, as the SDK would send it. */
    public static function machineName(): string
    {
        $ffi = Native::get();
        $buf = $ffi->new('char[256]');
        $r = $ffi->hlm_ffi_machine_name($buf, 256);
        if ($r !== 0) {
            throw new LicenseException($r);
        }

        return \FFI::string($buf);
    }

    /**
     * Offline one-shot: verify a signed license string and report its status
     * at $now (default: the system clock). Throws LicenseException when the
     * string is tampered or malformed.
     */
    public static function verify(
        string $jws,
        string $jwksJson,
        ?string $productId = null,
        ?string $machineId = null,
        ?\DateTimeInterface $now = null,
    ): LicenseStatus {
        $ffi = Native::get();
        $statusOut = $ffi->new('int');
        $r = $ffi->hlm_ffi_verify(
            $jws,
            $jwksJson,
            $productId,
            $machineId,
            $now?->getTimestamp() ?? 0,
            \FFI::addr($statusOut)
        );
        if ($r !== 0) {
            throw new LicenseException($r);
        }

        return LicenseStatus::from($statusOut->cdata);
    }

    /** Unix seconds -> DateTimeImmutable (UTC); INT64_MIN means "none". */
    private static function fromUnix(int $seconds): ?\DateTimeImmutable
    {
        if ($seconds === PHP_INT_MIN) { // HLM_TIME_NONE
            return null;
        }

        return new \DateTimeImmutable('@' . $seconds);
    }
}
